<a href="https://colab.research.google.com/github/FilipeSquire/Statistics-Tools/blob/main/ProductAffinityFundamentals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

affinity Analysis or Market Basket Analysis finds patterns to determine connections between purchases
in order that stores can increase their cross-selling potential

 itemset -> group of items with 2+ items in it
 Transactions -> main data source of affinity analysis. Each transaction has an ID, and it contains an itemset
 Support -> determines how popular an itemset is.  Support = Freq(i1, i2) of certain item / N

 Confidence -> likelihood of certain items to be purchased together. P(X|Y) = Support (X,Y) / P(X)

 Lift -> how likely Y is purchased when X is purchased while controlling for how popular butter is.
         Lift range is [o, +infinity]. When lift = 1, X and Y are independent.
         When lift > 1 it means that butter is likely to be purchased together with bread
         Lift(X > Y) = Support(X,Y)/Support(X)*Support(Y) = P(X,Y)/P(X)*P(Y)

Conviction -> ratio of expected frequency that X occurs without Y
 Conviction(X > Y) = 1 - Support(X) / 1 - Confidence(X > Y)

---



Creating a dataset for our practice exercise

In [1]:
basket=[['bread','butter','wine','bananas','coffee','carrots'],
          ['tomatoes','onions','cheese','milk','potatoes'],
          ['beer','chips','asparagus','salsa','milk','apples'],
          ['olive oil','bread','butter','tomatoes','steak','carrots'],
          ['tomatoes','onions','chips','wine','ketchup','orange juice'],
          ['bread','butter','beer','chips','milk'],
          ['butter','tomatoes','carrots','coffee','sugar'],
          ['tomatoes','onions','cheese','milk','potatoes'],
          ['bread','butter','ketchup','coffee','chicken wings'],
          ['butter','beer','chips','asparagus','apples'],
          ['tomatoes','onion','beer','chips','milk','coffee']]

Lets create a function to do the calculations

In [4]:
def frequency_items (x,y):
    fx=sum([x in i for i in basket])
    fy=sum([y in i for i in basket])
    
    fxy = sum([all(z in i for z in [x,y]) for i in basket])
    
    support = fxy/len(basket)
    confidence = support/(fx/len(basket))
    lift = confidence / (fy/len(basket))
    
    if confidence ==1:
        conviction = 0
    else:
        conviction=(1-(fy/len(basket)))/(1-confidence)
        
    print('Suppport = {}'.format(round(support,2)))
    print('Confidence = {}'.format(round(confidence,2)))
    print('Lift = {}'.format(round(lift,2)))
    print('Conviction= {}'.format(round(conviction,2)))
    
    print(fxy)

In [5]:
frequency_items('bread','butter')

Suppport = 0.36
Confidence = 1.0
Lift = 1.83
Conviction= 0
4


Reading it: 

36% of transactions contain both bread and butter.

Confidence indicates that butter is always purchased whenever bread is purchased

Lift > 1 means it's more likely that bread and butter will be bought together than individually
